cmd/common/instances.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/columns"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/columns/ellipsis"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/columns/formatter/textcolumns"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	grpcruntime "github.com/inspektor-gadget/inspektor-gadget/pkg/runtime/grpc"
+)
+
+type GadgetInfo struct {
+	pg *api.GadgetInstance
+}
+
+func AddInstanceCommands(
+	rootCmd *cobra.Command,
+	runtime *grpcruntime.Runtime,
+) {
+	runtimeParams := runtime.ParamDescs().ToParams()
+
+	findGadgetInstance := func(idOrName string) (*api.GadgetInstance, error) {
+		gadgetInstances, err := runtime.GetGadgetInstances(context.Background(), runtimeParams)
+		if err != nil {
+			return nil, err
+		}
+		var gadgetInstance *api.GadgetInstance
+		for _, tmpGadgetInstance := range gadgetInstances {
+			// Some heuristic to find the gadget the user requested
+
+			// Check name match first
+			if tmpGadgetInstance.Name == idOrName {
+				gadgetInstance = tmpGadgetInstance
+				break
+			}
+
+			// Check full match on ID
+			if tmpGadgetInstance.Id == idOrName {
+				gadgetInstance = tmpGadgetInstance
+				break
+			}
+
+			if len(idOrName) < 32 {
+				// allow partial matches
+				if strings.HasPrefix(tmpGadgetInstance.Id, idOrName) {
+					gadgetInstance = tmpGadgetInstance
+					break
+				}
+			}
+		}
+		if gadgetInstance == nil {
+			return nil, fmt.Errorf("not found")
+		}
+		return gadgetInstance, nil
+	}
+
+	var showDetails bool
+	listCmd := &cobra.Command{
+		Use:          "list",
+		Aliases:      []string{"l"},
+		SilenceUsage: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cols := columns.MustCreateColumns[GadgetInfo]()
```

Comment:
Can list command show the state of the gadget? Or is that even important for user? like can I attach to stopped gadget?

Reply:
Good question - initially I planned a way to stop/pause gadgets, which basically would mean to keep the configuration on the server(s), just not run the gadgets, so you can easily toggle - that's not yet implemented, though. I'm not sure if that is really needed and you should just "uninstall" and "reinstall" in that case.

Reply:
Also, "stopped" could be a state where the final data could remain buffered but still accessible.
_But_, and that's another UX thing: if we add that "stopped" state, the results/events would need to be stored to survive for example a restart of the gadget daemon.

Reply:
Do we need to worry about pause now? What's the use case for that? I can restart a pause gadget? I think that's a major change and it'll better if we can avoid that on this PR.